public/app/core/history/richHistoryLocalStorage.ts

```diff
@@ -0,0 +1,172 @@
+import RichHistoryStorage, {
+  RichHistoryServiceError,
+  RichHistoryStorageWarning,
+  RichHistoryStorageWarningDetails,
+} from './richHistoryStorage';
+import { RichHistoryQuery } from '../../types';
+import store from '../store';
+import { DataQuery } from '@grafana/data';
+import { isEqual, omit } from 'lodash';
+import { createRetentionPeriodBoundary, RICH_HISTORY_SETTING_KEYS } from './richHistoryLocalStorageUtils';
+
+export const RICH_HISTORY_KEY = 'grafana.explore.richHistory';
+export const MAX_HISTORY_ITEMS = 10000;
+
+/**
+ * Local storage implementation for Rich History. It keeps all entries in browser's local storage.
+ */
+export default class RichHistoryLocalStorage implements RichHistoryStorage {
+  /**
+   * Return all history entries, perform migration and clean up entries not matching retention policy.
+   */
+  async getRichHistory(): Promise<RichHistoryQuery[]> {
+    const richHistory: RichHistoryQuery[] = store.getObject(RICH_HISTORY_KEY, []);
+    const transformedRichHistory = migrateRichHistory(richHistory);
+    return transformedRichHistory;
+  }
+
+  async addToRichHistory(richHistoryQuery: RichHistoryQuery): Promise<RichHistoryStorageWarningDetails | undefined> {
+    let richHistory = cleanUp(await this.getRichHistory());
```

Comment:
```suggestion
    const richHistory = cleanUp(await this.getRichHistory());
```

Reply:
Hmm... maybe it'd be good to add it eslint rules?